submodules/wizard/wizard.js

```diff
@@ -2070,32 +2076,48 @@ define(function(require) {
 				callback = args.callback;
 
 			monster.waterfall([
-				function(waterfallCallback) {
-					if (appRestrictions.accessLevel === 'full') {
-						return waterfallCallback(null, []);
+				function getAllowedAppIds(waterfallCallback) {
+					if (appRestrictions.accessLevel === 'restricted') {
+						return waterfallCallback(null, appRestrictions.allowedAppIds);
 					}
 
+					var parentAccountId = self.wizardGetStore('parentAccountId'),
+						appsAccountId = self.wizardGetStore('resellerAccountId', parentAccountId);
+
+					// List all apps that the reseller or parent account has enabled
 					self.wizardGetAppList({
+						accountId: appsAccountId,
 						success: function(appList) {
-							waterfallCallback(null, appList);
+							waterfallCallback(null, _.map(appList, 'id'));
 						},
 						error: function(err) {
 							waterfallCallback(err);
 						}
 					});
 				},
-				function(appList, waterfallCallback) {
-					var blacklist = _
-						.chain(appList)
-						.map('id')
-						.difference(appRestrictions.allowedAppIds)
-						.value();
+				function getDefaultAppList(allowedAppIds, waterfallCallback) {
+					self.wizardGetAppList({
+						accountId: accountId,
+						success: function(appList) {
+							waterfallCallback(null, allowedAppIds, _.map(appList, 'id'));
+						},
```

Comment:
Looks like `getAllowedAppIds` and `getDefaultAppList` do not rely on each other. Therefore they can be started in parallel.

As it might create several layers of nested tasks, don't hesitate to extract them as function expressions and compose them, such as:

```js
var getAllowedAppsIds = function(next) {},
    getDefaultAppIds = function(next) {},
    getBlacklistAppIds = function(next) {
        monster.parallel({
            allowed: getAllowedAppsIds,
            defaults: getDefaultAppList
        }, function(err, result) {
            next(err, _.difference(result.defaults, result.allowed));
        });
    };

monster.waterfall([
    getBlacklistAppIds,
    function saveAppBlacklist(next) {}
], callback);
```

A good rule of thumb on when to do this is when you start struggling to name an async callbacks.